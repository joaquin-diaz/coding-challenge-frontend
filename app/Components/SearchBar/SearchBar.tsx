import * as React from 'react';
import styled from '@emotion/styled';

import Card from '../Card';
import Input from '../Input';
import Button from '../Button';

const { useState } = React;

const Container = styled(Card)`
  position: absolute;
  width: 40%;
  height: 50px;
  z-index: 10;
  left: 30px;
  top: 30px;
  display: flex;
  align-items: center;
  padding-left: 25px;
`;

const InputContainer = styled.div`
  flex: 0.7;
`;

const SearchButton = styled(Button)`
  position: absolute;
  right: -100px;
  height: 50px;
  font-weight: bold;
  padding: 15px;
`;

const SearchBar: React.FC = () => {
  const [searchValue, setSearchValue] = useState<string>('');

  return (
    <div>
      <Container>
        <InputContainer>
          <Input
            value={searchValue}
            onChange={setSearchValue}
            placeholder="Search by title, release year or main actor"
          />
        </InputContainer>
        <SearchButton>SEARCH</SearchButton>
      </Container>
    </div>
  );
};

export default SearchBar;
