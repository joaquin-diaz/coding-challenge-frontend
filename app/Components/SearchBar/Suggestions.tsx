import * as React from 'react';
import { consolidateSuggestions, Film } from '../../types/Film';
import Card from '../../Components/Card';
import styled from '@emotion/styled';
import styles from '../../constants/styles';

const Container = styled(Card)`
  width: 100%;
  padding: 10px 0;
`;

const Suggestion = styled.div`
  padding: 10px 15px;
  cursor: pointer;

  &:hover {
    background-color: ${styles.colors.gray};
  }
`;

const Match = styled.div`
  font-size: 12px;
  font-style: italic;
`;

const { useState, useEffect } = React;

interface SuggestionsProps {
  suggestedFilms: Array<Film>;
  query: string;
  onSuggestionClick: (title: string) => void;
}

const Suggestions: React.FC<SuggestionsProps> = props => {
  const [suggestions, setSuggestions] = useState<Array<string>>([]);
  useEffect(() => {
    const s = consolidateSuggestions(props.suggestedFilms, props.query);
    setSuggestions(s);
  }, [props.suggestedFilms.length]);

  const renderSuggestion = (suggestion: string, key: number) => {
    const [title, match] = suggestion.split('//');
    const handleClick = () => {
      props.onSuggestionClick(title);
    };

    return (
      <Suggestion key={key} onClick={handleClick}>
        <div>{title}</div>
        <Match>{match}</Match>
      </Suggestion>
    );
  };

  if (suggestions.length === 0) return null;

  return <Container>{suggestions.map(renderSuggestion)}</Container>;
};

export default Suggestions;
