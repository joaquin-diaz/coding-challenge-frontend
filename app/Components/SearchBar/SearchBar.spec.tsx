import * as React from 'react';

import SearchBar from './SearchBar';
import { shallow } from 'enzyme';

import debounce from 'lodash/debounce';
jest.mock('lodash/debounce', () => jest.fn(fn => fn));

const noop = () => {};

describe('<SearchBar />', () => {
  it('should render', () => {
    const wrapper = shallow(
      <SearchBar
        isLoading={false}
        isLoadingSuggestions={false}
        query="Something"
        onFetchSuggestions={noop}
        onSubmitSearch={noop}
        onQueryChange={noop}
        clearSuggestions={noop}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle input change', () => {
    (debounce as any).mockClear();
    const onChangeMock = jest.fn();
    const onSuggestionChange = jest.fn();

    const wrapper = shallow(
      <SearchBar
        isLoading={false}
        isLoadingSuggestions={false}
        query="Something"
        onFetchSuggestions={onSuggestionChange}
        onSubmitSearch={noop}
        onQueryChange={onChangeMock}
        clearSuggestions={noop}
      />,
    );

    const input = wrapper.find('Input');
    const inputOnChange = input.prop('onChange') as any;
    inputOnChange('something');

    expect(input.exists).toBeTruthy();
    expect(onChangeMock).toBeCalledWith('something');
    expect(onSuggestionChange).toBeCalledWith('something');
  });

  it('should handle submit', () => {
    const mockSubmit = jest.fn();
    const mockClearSuggestions = jest.fn();
    const mockPreventDefault = jest.fn();

    const wrapper = shallow(
      <SearchBar
        isLoading={false}
        isLoadingSuggestions={false}
        query="Something"
        onFetchSuggestions={noop}
        onSubmitSearch={mockSubmit}
        onQueryChange={noop}
        clearSuggestions={mockClearSuggestions}
      />,
    );

    const form = wrapper.find('form');
    expect(form.exists).toBeTruthy();
    form.simulate('submit', { preventDefault: mockPreventDefault });

    expect(mockSubmit).toBeCalledWith('Something');
    expect(mockPreventDefault).toBeCalled();
    expect(mockClearSuggestions).toBeCalled();
  });

  it('should be loading', () => {
    const wrapper = shallow(
      <SearchBar
        isLoading={true}
        isLoadingSuggestions={true}
        query="Something"
        onFetchSuggestions={noop}
        onSubmitSearch={noop}
        onQueryChange={noop}
        clearSuggestions={noop}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
