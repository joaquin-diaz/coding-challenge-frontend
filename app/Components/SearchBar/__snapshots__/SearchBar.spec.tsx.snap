// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> should be loading 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <Styled(div)>
      <Styled(Spinner)
        color="#3f51b5"
      />
      <Styled(div)>
        <Input
          onChange={[Function]}
          placeholder="Search by title, release year or main actor"
          value="Something"
        />
      </Styled(div)>
      <Styled(Button)
        disabled={true}
        type="submit"
      >
        SEARCH
      </Styled(Button)>
    </Styled(div)>
  </form>
  <Styled(div)>
    <SuggestionsContainer />
  </Styled(div)>
</Fragment>
`;

exports[`<SearchBar /> should render 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <Styled(div)>
      <Styled(div)>
        <Input
          onChange={[Function]}
          placeholder="Search by title, release year or main actor"
          value="Something"
        />
      </Styled(div)>
      <Styled(Button)
        disabled={false}
        type="submit"
      >
        SEARCH
      </Styled(Button)>
    </Styled(div)>
  </form>
  <Styled(div)>
    <SuggestionsContainer />
  </Styled(div)>
</Fragment>
`;
